Add YAMLs for gcp broker

liu-cong opened the pull request:
Fixes #808

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- Deploy broker data plane with static yamls

If `broker-targets` confimap doesn't exist, it will be created by the broker controller. Based on my experience, the controller usually starts faster than the data plane components, and the configmap is available when data plane components starts up. Sometimes if the controllers starts up slower, we may need to wait a few seconds for data plane pods to try the remount. But I have never seen it taking longer than a few seconds. 




<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- 🎁 Add new feature
- 🐛 Fix bug
- 🧽 Update or clean up current behavior
- 🗑️ Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
🗒️ If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note
Add deployment yamls for GCP broker data plane components.
```

**Docs**

<!--
📖 If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

/assign @grantr

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+# Copyright 2019 Google LLC
```
2020

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+  replicas: 1
+  selector:
+    matchLabels: &labels
+      app: cloud-run-events
+      role: broker-ingress
```
Maybe do similar things as https://github.com/knative/eventing/blob/master/pkg/reconciler/broker/resources/ingress.go#L155-L160

liu-cong replied:
I am using the same style as other deployments in `knative-gcp`. I think we should be consistent with `knative-gcp`. What do you tink?

yolocs replied:
I didn't realize that. I agree keeping it consistent at the moment is a better idea.

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+        volumeMounts:
+        - mountPath: /var/run/cloud-run-events/broker
+          name: broker-config
+        resources: # TODO(issue #876): determine good values for resource requests/limits
```
Do we need parameters for metrics/tracing/logging?

liu-cong replied:
No, not at the moment. I heard that eventing may stop supporting the configmap feature for metrics/tracing/logging. Need to follow up on that.

yolocs replied:
Hmm. What's the replacement? Operator?

liu-cong replied:
I didn't follow up on that. I heart from @grantr .

Maybe just stop supporting dynamic config.

grantr replied:
This was a conversation related to the dissolution of the Knative Observability working group. Evan questioned if the complexity of the dynamic configuration was worth maintaining. No concrete guidance yet.

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+    metadata:
+      labels: *labels
+    spec:
+      containers:
```
Since ingress is a service, better to have readiness and liveness probes?

liu-cong replied:
We will be serving those zpages in https://github.com/google/knative-gcp/issues/785, also created an issue to add the probes:https://github.com/google/knative-gcp/issues/882

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+  replicas: 1
+  selector:
+    matchLabels: &labels
+      app: cloud-run-events
+      role: broker-fanout
```
Similarly, maybe we should do more knative-style label keys

liu-cong replied:
Same as https://github.com/google/knative-gcp/pull/880#discussion_r410309018

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+      labels: *labels
+    spec:
+      containers:
+      - name: broker-ingress
```
Does ingress need its own service account?

liu-cong replied:
No, it doesn't at the moment.

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+  replicas: 1
+  selector:
+    matchLabels: &labels
+      app: cloud-run-events
+      role: broker-retry
```
Similar to my other two comments.

liu-cong replied:
Same as https://github.com/google/knative-gcp/pull/880#discussion_r410309018

yolocs commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,157 @@
+    metadata:
+      labels: *labels
+    spec:
+      containers:
```
Also specify `ports` here with name `http`. I believe some mesh (e.g. istio) requires proper port naming.

yolocs commented on config/core/services/broker.yaml:
```diff
@@ -0,0 +1,30 @@
+    app: cloud-run-events
+    role: broker-ingress
+  ports:
+    - name: ingress
```
Better using `http` as the name.

liu-cong commented:
/retest

liu-cong commented:
/retest

liu-cong commented:
/retest

liu-cong commented:
/retest

liu-cong commented:
/hold 

Should support both secret and workload identity

liu-cong commented on config/core/deployments/controller.yaml:
```diff
@@ -78,5 +78,4 @@ spec:
         secret:
           secretName: google-cloud-key
           optional: true
-      serviceAccount: controller
```
We already have `serviceAccountName` specified. `serviceAccount` is deprecated in favor of `serviceAccountName`

grantr commented on config/core/deployments/broker.yaml:
```diff
@@ -0,0 +1,190 @@
+      containers:
+      - name: broker-ingress
+        image: ko://github.com/google/knative-gcp/cmd/broker/ingress
+        env:
```
The other two deployments have `SYSTEM_NAMESPACE`, `POD_NAME`, and metrics-related env vars. Why does the ingress container not have them? Seems like we'll need them when we start adding metrics and tracing.

liu-cong replied:
The other two currently reports memstats and therefore need metrics domain env var. 

Yes once we add metrics to ingress we will need those env vars as well.  I can add those now or wait until we have metrics added.

grantr replied:
Either way is fine with me. Feel free to unhold if you'd like to add the env vars later.

liu-cong replied:
Sounds good. The binary should fail to start if we forget to add the required env var. So we we will find out easily if we forget to add them

liu-cong commented:
/unhold

liu-cong commented:
@grantr @yolocs 

I made some updates and now it's ready for review:
- Use the `cre-pubsub` GSA instead of the `cloud-run-events` GSA which is used by the controller.
- Moved broker yamls to `config/broker` so they are not installed by default. Currently GCP broker is an optional installation.
- Added an initial doc for how to install the GCP broker.

liu-cong commented:
/retest

liu-cong commented:
cc @capri-xiyue

capri-xiyue commented:
I tested the installation guide with secret, it works.
